### PR TITLE
Bug 1920320: Show link to common template in vm details view

### PR DESCRIPTION
**Pull request by yaacov**
We now consider common templates as first class template citizens, we need to link to them from details page.

Screenshot:
Before:
screenshot-localhost_9000-2021 01 27-17_30_49

After:
screenshot-localhost_9000-2021 01 27-17_30_15

### CODE REVIEW

File: frontend/packages/kubevirt-plugin/src/components/vms/vm-resource.tsx [hunk 1]
```diff
@@ -91,11 +101,25 @@ export const VMResourceSummary: React.FC<VMResourceSummaryProps> = ({
   kindObj,
 }) => {
   const { t } = useTranslation();
+  const templatesResource: WatchK8sResource = {
```
Comment (rawagner) [inline frontend/packages/kubevirt-plugin/src/components/vms/vm-resource.tsx:104]: We should load the templates together with user templates at `vm-details-page`

File: frontend/packages/kubevirt-plugin/src/components/vms/vm-resource.tsx [hunk 2]
```diff
@@ -91,11 +101,25 @@ export const VMResourceSummary: React.FC<VMResourceSummaryProps> = ({
...
 
   const isVM = kindObj === VirtualMachineModel;
   const vmiLike = isVM ? vm : vmi;
 
-  const template = getVMTemplate(vm, templates);
+  const isCommonTemplate =
+    getLabel(vmiLike, LABEL_USED_TEMPLATE_NAMESPACE) === TEMPLATE_VM_COMMON_NAMESPACE;
+  const template = isCommonTemplate
+    ? getVMCommonTemplate(vmiLike, commonTemplates)
+    : getVMTemplate(vmiLike, templates);
```
Comment (rawagner) [inline frontend/packages/kubevirt-plugin/src/components/vms/vm-resource.tsx:122]: this change is not needed when you fetch common templates with user templates

**Review by rawagner — CHANGES_REQUESTED** [review]

File: frontend/packages/kubevirt-plugin/src/components/vms/vm-resource.tsx (hunk 1 above)
Comment (yaacov) [reply frontend/packages/kubevirt-plugin/src/components/vms/vm-resource.tsx]: removed the `WatchK8sResource` hook and now using the firehose mechanism for both `templats` and `commonTemplates`

File: frontend/packages/kubevirt-plugin/src/components/vms/vm-resource.tsx (hunk 2 above)
Comment (yaacov) [reply frontend/packages/kubevirt-plugin/src/components/vms/vm-resource.tsx:104]: in the testing environment we have VMs that do not have the template name label when created from common template, so the common template can't use the same method, using the  `getVMCommonTemplate` to catch this cases.

**Comment by yaacov** [comment]
@rawagner please re review

File: frontend/packages/kubevirt-plugin/src/selectors/vm-template/selectors.ts [hunk 3]
```diff
@@ -26,6 +33,25 @@ export const getVMTemplate = (
     : undefined;
 };
 
+export const getVMCommonTemplate = (
```
Comment (rawagner) [inline frontend/packages/kubevirt-plugin/src/selectors/vm-template/selectors.ts:36]: you should still find the template by name and namespace. If these name/namespace labels are missing than you can fallback to label/workload/os.

**Review by rawagner — CHANGES_REQUESTED** [review]

File: frontend/packages/kubevirt-plugin/src/selectors/vm-template/selectors.ts (hunk 3 above)
Comment (yaacov) [reply frontend/packages/kubevirt-plugin/src/selectors/vm-template/selectors.ts:36]: added the regular check first

**Comment by yaacov** [comment]
@rawagner please re review

**Comment by spadgett** [comment]
/override ci/prow/ceph-storage-plugin

The ceph-storage-plugin job should only have run on PRs with changes to `frontend/packages/ceph-storage-plugin`. It failed due to an unrelated CI config change. Overriding status on this PR.

File: frontend/packages/kubevirt-plugin/src/components/vms/vmDetailsItemTemplate.tsx [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  const templatesResource: WatchK8sResource = {
+    isList: false,
+    optional: true,
+    kind: TemplateModel.kind,
+    name,
+    namespace,
+    selector: {
+      matchLabels: {
+        [TEMPLATE_TYPE_LABEL]:
+          namespace === TEMPLATE_VM_COMMON_NAMESPACE ? TEMPLATE_TYPE_BASE : TEMPLATE_TYPE_VM,
```
Comment (rawagner) [inline frontend/packages/kubevirt-plugin/src/components/vms/vmDetailsItemTemplate.tsx:30]: since you have name and namespace, there's no need to match by label too. BTW in theory you can have user template in openshift namespace too.

File: frontend/packages/kubevirt-plugin/src/components/vms/vmDetailsItemTemplate.tsx [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+} from '../../constants';
+import { getBasicID, prefixedID } from '../../utils';
+import { VMTemplateLink } from '../vm-templates/vm-template-link';
+import VMDetailsItem from './vmDetailsItem';
+
+const VMDetailsItemTemplate: React.FC<VMDetailsItemTemplateProps> = ({ name, namespace }) => {
+  const { t } = useTranslation();
+  const templatesResource: WatchK8sResource = {
+    isList: false,
+    optional: true,
```
Comment (rawagner) [inline frontend/packages/kubevirt-plugin/src/components/vms/vmDetailsItemTemplate.tsx:23]: no need for optional. The hook doesnt really care.

File: frontend/packages/kubevirt-plugin/src/components/vms/vmDetailsItemTemplate.tsx [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  const [template, loadedTemplates, errorTemplates] = useK8sWatchResource<TemplateKind>(
+    templatesResource,
+  );
+  const id = getBasicID(template);
+
+  return (
+    <VMDetailsItem
+      title={t('kubevirt-plugin~Template')}
+      idValue={prefixedID(id, 'template')}
+      isNotAvail={!name || !namespace || !loadedTemplates || errorTemplates}
```
Comment (rawagner) [inline frontend/packages/kubevirt-plugin/src/components/vms/vmDetailsItemTemplate.tsx:43]: you need some loading state. Showing `Not available` if `!loadedTemplates` is not good

**Review by rawagner — CHANGES_REQUESTED** [review]

File: frontend/packages/kubevirt-plugin/src/components/vms/VMDetailsItem.tsx [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+      </dt>
+      <dd id={idValue} className={valueClassName}>
+        {isNotAvail && (
+          <span className="text-secondary">
+            {isNotAvailMessage || t('kubevirt-plugin~Not available')}
+          </span>
+        )}
+        {isLoading && !isNotAvail && <LoadingInline />}
+        {!isNotAvail && !isLoading && children}
+      </dd>
```
Comment (rawagner) [inline frontend/packages/kubevirt-plugin/src/components/vms/VMDetailsItem.tsx:46]: I find these conditions hard to follow. Can you do something like

```suggestion
if (isNotAvail) {
  body = (
    <span className="text-secondary">
      {isNotAvailMessage || t('kubevirt-plugin~Not available')}
    </span>
  );
} else if (isLoading) {
  body = <LoadingInline />;
} else {
  body = children;
}



....
      <dd id={idValue} className={valueClassName}>
         {body}
      </dd>
```

File: frontend/packages/kubevirt-plugin/src/components/vms/VMDetailsItemTemplate.tsx [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    templatesResource,
+  );
+  const id = getBasicID(template);
+
+  return (
+    <VMDetailsItem
+      title={t('kubevirt-plugin~Template')}
+      idValue={prefixedID(id, 'template')}
+      isLoading={!loadedTemplates}
+      isNotAvail={!name || !namespace || !template || errorTemplates}
```
Comment (rawagner) [inline frontend/packages/kubevirt-plugin/src/components/vms/VMDetailsItemTemplate.tsx:31]:
```suggestion
      isNotAvail={!name || !namespace || (loadedTemplates && !template) || errorTemplates}
```

**Review by rawagner — CHANGES_REQUESTED** [review]
